File: src/dfm/hooks.py
"""Hook management for Profiles."""

import logging
import shlex
import subprocess
import sys

logger = logging.getLogger(__name__)


class Hooks:
    """Manages and executes hooks."""

    def __init__(self, where, hooks=None):
        self.where = where
        self.hooks = hooks if hooks is not None else {}

    @classmethod
    def from_config(cls, where, config):
        """Load Hooks from a .dfm.yml config file."""
        return cls(where=where, hooks=config.pop("hooks", {}))

    def run_hook(self, name, dry_run=False):
        """
        Run the hook with name.

        If dry_run is given the command will not be actually executed.
        """
        commands = self.hooks.get(name, [])
        for command in commands:
            if isinstance(command, dict):
                interpreter = shlex.split(command.get("interpreter", "/bin/sh -c"))
                script = command.get("script", "")
            else:
                interpreter = ["/bin/sh", "-c"]
                script = command

            if not script:
                logger.warning(
                    "Found an empty script for hook: %s, skipping",
                    name,
                )
                continue

            if dry_run:
                logger.debug("Running hook %s command: %s", name, command)
                continue

            try:
                subprocess.call(
                    interpreter + [script],
                    cwd=self.where,
                    stdin=sys.stdin,
                    stdout=sys.stdout,
                    stderr=sys.stderr,
                )
            except subprocess.CalledProcessError as proc_err:
                logger.error(
                    "command %s exited with non-zero error: %s",
                    command,
                    proc_err,
                )

File: profiles/hooks.go
package profiles

import (
	"fmt"
	"os"
	"os/exec"

	"github.com/chasinglogic/dfm/logger"
	"github.com/google/shlex"
)

type Hooks map[string][]interface{}

func (h Hooks) RunHook(name, dir string, dryRun bool) error {
	hooks, ok := h[name]
	if !ok {
		logger.Debug.Printf("no hook defined for: %s", name)
		return nil
	}

	for _, hook := range hooks {
		var command []string
		var err error

		if hookStr, ok := hook.(string); ok {
			command = []string{"sh", "-c", hookStr}
		} else if hookMap, ok := hook.(map[string]interface{}); ok {
			interpreter, hasInterpreter := hookMap["interpreter"]
			script, hasScript := hookMap["script"]
			if !hasScript || !hasInterpreter {
				return fmt.Errorf("%s hook is missing script or interpreter", name)
			}

			if _, ok := interpreter.(string); !ok {
				return fmt.Errorf("%s hook interpreter is not a string", name)
			}

			if _, ok := script.(string); !ok {
				return fmt.Errorf("%s hook script is not a string", name)
			}

			interpreterArgs, err := shlex.Split(interpreter.(string))
			if err != nil {
				return err
			}

			command = append(
				interpreterArgs,
				script.(string),
			)
		}

		logger.Debug.Printf("Executing command: %s\n", command)
		proc := exec.Command(command[0], command[1:]...)
		proc.Stdout = os.Stdout
		proc.Stdin = os.Stdin
		proc.Stderr = os.Stderr
		proc.Dir = dir

		proc.Start()
		err = proc.Wait()

		if err != nil {
			return err
		}
	}

	return nil
}
